src/main/java/seedu/modtrek/model/DegreeProgressionData.java

```diff
@@ -0,0 +1,74 @@
+package seedu.modtrek.model;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import seedu.modtrek.model.module.Module;
+import seedu.modtrek.model.module.UniqueModuleList;
+
+/**
+ * DegreeProgressionData holds all the data for an overview of Degree Progression.
+ */
+public class DegreeProgressionData {
+
+    // Currently for cohort 2122
+    public static final int TOTALCREDIT = 160;
+    public static final Map<String, Integer> REQUIREMENTCREDITS = Map.of(
+            "UNIVERSITY_LEVEL_REQUIREMENTS", 16,
+            "COMPUTER_SCIENCE_FOUNDATION", 36,
+            "COMPUTER_SCIENCE_BREADTH_AND_DEPTH", 40,
+            "IT_PROFESSIONALISM", 12,
+            "MATHEMATICS_AND_SCIENCES", 16,
+            "UNRESTRICTED_ELECTIVES", 40);
+    private int currentCredit = 0;
+    private HashMap<String, Integer> completedRequirementCredits = new HashMap<>();
+    private float cumulativePoints = 0;
+    private float gpa;
+
+    private DegreeProgressionData() {}
+
+    /**
+     * Factory method to generate the data needed to show progress. Progress
+     * is calculated from the user's current modlist.
+     *
+     * @param modList
+     * @return
+     */
+    public static DegreeProgressionData generate(UniqueModuleList modList) {
+        DegreeProgressionData data = new DegreeProgressionData();
+        modList.forEach((module) -> {
+            data.computeModule(module);
+        });
+        data.computeGpa();
+        return data;
+    }
+
+    public int getCurrentCredit() {
+        return currentCredit;
+    }
+
+    public Map<String, Integer> getCompletedRequirementCredits() {
+        return completedRequirementCredits;
+    }
+
+    public float getGpa() {
+        return gpa;
+    }
+
+    private void computeModule(Module module) {
+        int credit = Integer.valueOf(module.getCredit().toString());
+        module.getTags().forEach((tag) -> {
+            completedRequirementCredits.merge(tag.toString(),
+                    credit, (oldValue, newValue) -> {
+                        return oldValue + newValue;
+                    });
+        });
+        currentCredit += credit;
```

Comment:
Actually, I think it is possible in some circumstances for one module to be double counted. E.g. CS1101S has 4MCs going to ULR and 4MCs going to CSF. In the end if I'm not wrong, our CAP should takes into account CS1101S grade for 8MCs.

If I'm right, then we should move line 66 into the `forEach` loop:
```
module.getTags().forEach((tag) -> {
    completedRequirementCredits.merge(tag.toString(),
    credit, (oldValue, newValue) -> {
        return oldValue + newValue;
    });
    currentCredit += credit;
});
```

Reply:
Hmm... I don't think the module will be counted for 8 MC. If I recall correctly, 1101S will map to both ULR and CSF, hence increasing the progression for both areas (I +4 to both tags here), but overall should be a +4MCs. Technically, the UE cap here should +4 too so that total MCs goes up to 160 with the double counting (I didn't increase the cap for UE here yet). Not sure if this applies to all other double mapping mods...

Reply:
Oh, I think the UE cap doesn't have to increase. According to the curriculum [here](https://www.comp.nus.edu.sg/cugresource/per-cohort/cs/cs-21-22/#_ftn2), the 40MCs of UE already takes into account the double counting of modules (e.g. CS1101S, ES2660).

Reply:
Wait does it? If I add CS1101S as 4 MC into ULR and another 4 into CSF, then by right I need to count 8 MC into GPA. But I think they only count 4 in total since they disallow double counting, which means 4 MC allocated to UE to ensure 160MCs is still done in total?

Reply:
Yea, they already calculated the number of UEs for us to take into account the double counting.